src/os/bluestore/BlueStore.cc

```diff
@@ -3998,14 +3998,19 @@ int BlueStore::_decompress(bufferlist& source, bufferlist* result)
   bluestore_compression_header_t chdr;
   ::decode(chdr, i);
   string name = bluestore_blob_t::get_comp_alg_name(chdr.type);
-  CompressorRef compressor = Compressor::create(cct, name);
-  if (!compressor.get()) {
+  CompressorRef cp;
+  if (compressor != nullptr && compressor->get_type() == name)
```

Comment:
Could you please do an assignment cp = compressor; first to avoid potential race condition with _set_compression method?